### PR TITLE
feat(preheat):add preheat api, controller and manager

### PR DESCRIPTION
**Pull request by kofj**
- define instance's api
- define extension models for api
- implement preheat controller
- implement preheat manager
- most code are picked up from the original P2P feat branch

Signed-off-by: fanjiankong <fanjiankong@tencent.com>

### CODE REVIEW

File: src/pkg/p2p/preheat/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// extURL, err := config.ExtURL()
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	access := []*tk.ResourceActions{
```
Comment (steven-zou) [inline src/pkg/p2p/preheat/controller/controller.go:253]: here we may need more discussions. Similar to the scanning, the pull action may be blocked by the security settings (set at the project configurations, such as signature and vulnerability severity)

File: src/pkg/p2p/preheat/dao/instance_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
+package dao
```
Comment (steven-zou) [inline src/pkg/p2p/preheat/dao/instance_test.go:1]: Provider instance DAO and manager seem to be @chlins' work.  @chlins, could you please take a detailed review of such parts?

File: api/v2.0/swagger.yaml [hunk 3]
```diff
@@ -1098,3 +1251,87 @@ definitions:
...
+      version:
+        type: string
+        description: version
+      source:
+        type: string
+        description: source
+      auth_mode:
+        type: string
+        description: auth mode
+  Instance:
```
[the remark this answers is not in the document]
Comment (chlins) [reply api/v2.0/swagger.yaml:1280]: The instance definitions looks old, needs to sync with provider instance model in master branch.

File: src/pkg/p2p/preheat/models/extensions.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
+package models
```
Comment (chlins) [inline src/pkg/p2p/preheat/models/extensions.go:1]: If we do not need this file now, maybe we can delete it.

File: api/v2.0/swagger.yaml [hunk 5]
```diff
@@ -1098,3 +1251,87 @@ definitions:
...
+        type: string
+        description: The health status
+      enabled:
+        type: boolean
+        description: Whether the instance is activated or not
+      setup_timestamp:
+        type: integer
+        format: int64
+        description: The timestamp of instance setting up
+      extensions:
```
Comment (chlins) [inline api/v2.0/swagger.yaml:1316]: extensions is removed in new model.

File: api/v2.0/swagger.yaml [hunk 6]
```diff
@@ -1098,3 +1251,87 @@ definitions:
...
+        description: The authentication way supported
+      auth_data:
+        type: object
+        description: The auth credential data if exists
+        additionalProperties:
+          type: string
+      status:
+        type: string
+        description: The health status
+      enabled:
```
Comment (chlins) [inline api/v2.0/swagger.yaml:1309]: have a new properties: default(represents whether the instance is default)

File: src/pkg/p2p/preheat/models/extensions.go (hunk 4 above)
Comment (kofj) [reply src/pkg/p2p/preheat/models/extensions.go]: It's need by the old code from p2p branch, I'll refactoring it later.

File: api/v2.0/swagger.yaml [hunk 7]
```diff
@@ -641,6 +641,153 @@ paths:
           $ref: '#/responses/401'
         '500':
           $ref: '#/responses/500'
+  /distribution/instances:
+    get:
+      summary: List p2p instances
```
[the remark this answers is not in the document]
Comment (kofj) [reply api/v2.0/swagger.yaml:646]: updated

File: api/v2.0/swagger.yaml [hunk 8]
```diff
@@ -1098,3 +1251,87 @@ definitions:
       op_time:
         type: string
         description: The time when this operation is triggered.
+  Provider:
```
[the remark this answers is not in the document]
Comment (kofj) [reply api/v2.0/swagger.yaml:1254]: Yeah, I'l rename it to Metadata.

File: api/v2.0/swagger.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kofj) [reply api/v2.0/swagger.yaml:1280]: Synced.

File: api/v2.0/swagger.yaml [hunk 9]
```diff
@@ -641,6 +641,153 @@ paths:
...
+          $ref: '#/responses/401'
+        '403':
+          $ref: '#/responses/403'
+        '404':
+          $ref: '#/responses/404'
+        '500':
+          $ref: '#/responses/500'
+    put:
+      summary: Update instance
+      description: Update instance
```
[the remark this answers is not in the document]
Comment (kofj) [reply api/v2.0/swagger.yaml:760]: All swagger's description updated.

File: src/pkg/p2p/preheat/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
+package preheat
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/pkg/p2p/preheat/controller/controller.go]: Moved

File: api/v2.0/swagger.yaml (hunk 5 above)
Comment (kofj) [reply api/v2.0/swagger.yaml:1316]: Removed.

File: api/v2.0/swagger.yaml (hunk 6 above)
Comment (kofj) [reply api/v2.0/swagger.yaml:1309]: Added.

File: src/pkg/p2p/preheat/controller/controller.go (hunk 1 above)
Comment (kofj) [reply src/pkg/p2p/preheat/controller/controller.go]: Let us  discussions later.

File: src/pkg/p2p/preheat/instance/manager.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+	return &manager{
+		dao: dao.New(),
+	}
+}
+
+// Ensure *manager has implemented Manager interface.
+var _ Manager = (*manager)(nil)
+
+var (
+	errNilMetadataModel = errors.New("nil instance metadata model")
```
Comment (chlins) [inline src/pkg/p2p/preheat/instance/manager.go:80]: this variable maybe useless?

File: src/pkg/p2p/preheat/dao/instance/dao.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+// List lists instances by query params.
+func (i *dao) List(ctx context.Context, query *q.Query) (total int64, ins []*provider.Instance, err error) {
+	var o beego_orm.Ormer
+	o, err = orm.FromContext(ctx)
+	if err != nil {
+		return
+	}
+	qt := o.QueryTable(new(provider.Instance))
+
+	if query != nil {
```
Comment (steven-zou) [inline src/pkg/p2p/preheat/dao/instance/dao.go:93]: Here, we should follow the query setter approach used in other modules for keeping consistent.

File: src/pkg/p2p/preheat/instance/manager.go (hunk 11 above)
Comment (steven-zou) [reply src/pkg/p2p/preheat/instance/manager.go:80]: @kofj correct?

File: src/pkg/p2p/preheat/dao/instance/dao.go (hunk 12 above)
Comment (kofj) [reply src/pkg/p2p/preheat/dao/instance/dao.go:93]: Use "github.com/goharbor/harbor/src/lib/orm".QuerySetter method?

File: src/controller/p2p/preheat/controller.go [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+	if err != nil {
+		return 0, err
+	}
+	for _, theOne := range allOnes {
+		if theOne.Endpoint == instance.Endpoint {
+			return 0, ErrorConflict
+		}
+	}
+	// Check health before saving
+	f, ok := provider.GetProvider(instance.Vendor)
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/controller/p2p/preheat/controller.go:136]: That's good, I'll add health check API in the next PR. But, I don't think it's appropriate if we don't check the user's input parameters on the backend.

File: src/controller/p2p/preheat/controller.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+	if instance == nil {
+		return 0, errors.New("nil instance object provided")
+	}
+
+	// Avoid duplicated endpoint
+	_, allOnes, err := cc.iManager.List(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+	for _, theOne := range allOnes {
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/controller/p2p/preheat/controller.go:130]: Great!

File: src/controller/p2p/preheat/controller.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+	// Ctl is a global preheat controller instance
+	Ctl = NewController()
+)
+
+// ErrorConflict for handling conflicts
+var ErrorConflict = errors.New("resource conflict")
+
+// CompositePreheatingResults handle preheating results among multiple providers
+// Key is the ID of the provider instance.
+type CompositePreheatingResults map[int64]*[]*provider.PreheatingStatus
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/controller/p2p/preheat/controller.go:28]: OK

File: codecov.yml [hunk 16]
```diff
@@ -36,5 +36,5 @@ ignore:
   - "tests"
   - "tools"
   - "src/vendor"
-  - "src/server/v2.0/models/**/*"
-  - "src/server/v2.0/restapi/**/*"
+  - "src/github.com/goharbor/harbor/src/server/v2.0/restapi/**/"
```
[the remark this answers is not in the document]
Comment (kofj) [reply codecov.yml:39]: But it also works.

File: src/controller/p2p/preheat/controller.go [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+	return cc.iManager.Delete(ctx, id)
+}
+
+// Update implements @Controller.Update
+func (cc *controller) UpdateInstance(ctx context.Context, instance *providerModels.Instance, properties ...string) error {
+	if len(properties) == 0 {
+		return errors.New("no properties provided to update")
+	}
+
+	return cc.iManager.Update(ctx, instance, properties...)
```
[the remark this answers is not in the document]
Comment (kofj) [reply src/controller/p2p/preheat/controller.go:169]: OK, I'll check default instance first.

File: codecov.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (steven-zou) [reply codecov.yml:39]: What about other folders, can we keep them consistent with [codecov ](https://docs.codecov.io/docs/ignoring-paths)doc?

File: src/controller/p2p/preheat/controller.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	num, err := cc.iManager.Count(ctx, query)
+	if err != nil {
+		return 0, err
+	}
+	if num > 0 {
+		return 0, ErrorConflict
+	}
+
+	// !WARN: Check healthy status at fronted.
```
Comment (steven-zou) [inline src/controller/p2p/preheat/controller.go:141]: But don't forget to provide a new health check API endpoint later.

File: src/pkg/p2p/preheat/dao/instance/dao.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+	err = o.Begin()
+	if err != nil {
+		return err
+	}
+
+	// check default instances first
```
Comment (steven-zou) [inline src/pkg/p2p/preheat/dao/instance/dao.go:74]: default handler should happen at least in the mgr or controller layer, not DAO here. DAO only needs to handle the concrete SQL actions.

File: src/pkg/p2p/preheat/dao/instance/dao.go (hunk 12 above)
Comment (steven-zou) [reply src/pkg/p2p/preheat/dao/instance/dao.go:125]: yup

File: src/controller/p2p/preheat/controller.go (hunk 18 above)
Comment (kofj) [reply src/controller/p2p/preheat/controller.go:141]: Added TODO

**Pull request merged**